File: src/code_builder.js
/**
 * code
 */

const { KEYS, TYPES, CONFIGS } = require("./constants.js");
const { js, files, tools, encryption } = require("./tools.js");
const { MyBuilderBase } = require("./builder_base.js");

class MyCodeBuilder extends MyBuilderBase {

    constructor() {
        super("code");
        this._lineArr = [];
        this._macroStartTag = "[M[";
        this._macroEndTag = "]M]";
        this._commentTag = "//";
        this._headFormat = '{0} {1}';
    }

    setComment(commentTag, headFormat) {
        js.assert(js.is_text(commentTag), 'invalid comment tag');
        this._commentTag = commentTag;
        console.log("\n\n\n-->", "comment", commentTag, headFormat);
        if (headFormat !== undefined) {
            this._headFormat = headFormat;
        }
        return this;
    }

    onMacro(macroCallback) {
        this._onMacroCallback = macroCallback;
        return this;
    }

    onLine(lineCallback) {
        this._onLineCallback = lineCallback;
        return this;
    }

    _COMMAND_FILE_BASE64(code, args) {
        const filePath = args[0];
        this._assert(files.is_file(filePath), `file not found, path: ${filePath}`);
        const content = files.read(filePath, 'base64');
        return code.format(content);
    }

    _COMMAND_FILE_PLAIN(code, args) {
        const filePath = args[0];
        this._assert(files.is_file(filePath), `file not found, path: ${filePath}`);
        const content = files.read(filePath, 'utf8');
        return code.format(content);
    }

    _COMMAND_FILE_STRING(code, args) {
        let filePath = args[0];
        const escapeTag = args[1] || "";
        const minimize = args[2] !== undefined && args[2].toLowerCase() === "true";
        filePath = this._projDir + filePath;
        this._assert(files.is_file(filePath), `file not found, path: ${filePath}`);
        let fileContent = files.read(filePath);
        let lineArr = fileContent.split("\n");
        for (let i = 0; i < lineArr.length; i++) {
            lineArr[i] = lineArr[i]
                .replace(/[\n\r]+$/, " ")
                .replace(/\\/g, `${escapeTag}\\${escapeTag}\\`)
                .replace(/"/g, `${escapeTag}\\${escapeTag}"`)
                .replace(/'/g, `${escapeTag}\\${escapeTag}'`);
        }
        let result = lineArr.join(` ${escapeTag}\\n `);
        if (minimize) {
            result = result.replace(/\s+/g, " ");
        }
        js.print("\n\n\n--->", result);
        return code.format(result);
    }


    _COMMAND_LINE_REFPLACE(code, args) {
        return args[0];
    }

    _parseLine(index, line) {
        const commentPosition = line.indexOf(this._commentTag);
        if (commentPosition === -1) {
            if (this._onLineCallback) {
                return this._onLineCallback(line);
            }
            return line;
        }
        const macroStartIndex = line.indexOf(this._macroStartTag);
        const macroEndIndex = line.indexOf(this._macroEndTag);
        if (macroStartIndex === -1 || macroEndIndex === -1 || macroStartIndex >= macroEndIndex) {
            if (this._onLineCallback) {
                return this._onLineCallback(line);
            }
            return line;
        }
        const code = line.substring(0, commentPosition);
        const macro = line.substring(macroStartIndex + this._macroEndTag.length, macroEndIndex);
        const body = macro.split("|");
        this._assert(body.length >= 1, "invalid macro, line: " + line);
        const command = body[0].trim();
        const args = body.slice(1).map(arg => arg.trim());
        if (this['_COMMAND_' + command]) {
            return this['_COMMAND_' + command](code, args);
        } else if (this._onMacroCallback) {
            return this._onMacroCallback(code, command, ...args);
        } else {
            return line;
        }
    }

    _processBuild() {
        this._assert(!Array.isArray(this._inputFiles) || this._inputFiles.length > 0, "input files are not defined");
        this._assert(typeof this._commentTag === "string", "comment tag is not defined");

        this._print("reading files ...");
        for (let i = 0; i < this._inputFiles.length; i++) {
            const path = this._inputFiles[i];
            this._assert(files.is_file(path), "file not found: " + path);
            const content = files.read(path);
            this._assert(content.length > 0, "input files are empty");
            const lineArr = content.split("\n");
            //
            if (js.is_text(this._headFormat)) {
                const currName = this._inputNames[i];
                const currDate = new Date().toLocaleString();
                const headInfo = this._headFormat.format(currName, currDate);
                this._lineArr.push("");
                this._lineArr.push(this._commentTag + " " + headInfo);
                this._lineArr.push("");
            }
            //
            for (let index = 0; index < lineArr.length; index++) {
                const line = lineArr[index];
                const newLine = this._parseLine(index, line);
                if (Array.isArray(newLine)) {
                    for (let i = 0; i < newLine.length; i++) {
                        this._lineArr.push(newLine[i]);
                    }
                } else if (typeof newLine === "string") {
                    this._lineArr.push(newLine);
                }
            }
        }

        this._print("creating target ...");
        const html = this._lineArr.join("\n");
        this._assert(this._outputFile !== undefined, "output path not found");
        files.write(this._outputFile, html);
        this._print("writing target succeeded!");
        return this;
    }

}

module.exports = {MyCodeBuilder};

File: src/code_builder.py
import os
import base64
from datetime import datetime

from tools import *
from builder_base import *

class MyCodeBuilder(MyBuilderBase):

    def __init__(self, *args):
        super().__init__("code", *args)
        self._lineArr = []
        self._macroStartTag = "[M["
        self._macroEndTag = "]M]"
        self._commentTag = "//"
        self._headFormat = '{0} {1}'

    def setComment(self, commentTag, headFormat=None):
        py.check(py.is_text(commentTag), 'invalid comment tag')
        self._commentTag = commentTag
        if headFormat is not None:
            self._headFormat = headFormat
        return self

    def onMacro(self, macroCallback):
        self._onMacroCallback = macroCallback
        return self

    def onLine(self, lineCallback):
        self._onLineCallback = lineCallback
        return self

    def _COMMAND_FILE_BASE64(self, code, args):
        filePath = args[0]
        self._assert(files.is_file(filePath), f"file not found, path: {filePath}")
        with open(filePath, 'rb') as f:
            content = base64.b64encode(f.read()).decode('utf-8')
        return code.format(content)

    def _COMMAND_FILE_PLAIN(self, code, args):
        filePath = args[0]
        self._assert(files.is_file(filePath), f"file not found, path: {filePath}")
        with open(filePath, 'r', encoding='utf-8') as f:
            content = f.read()
        return code.format(content)

    def _COMMAND_FILE_STRING(self, code, args):
        filePath = args[0]
        escapeTag = args[1] if len(args) > 1 else ""
        minimize = len(args) > 2 and args[2].lower() == "true"
        filePath = os.path.join(self._projDir, filePath)
        self._assert(files.is_file(filePath), f"file not found, path: {filePath}")
        with open(filePath, 'r') as f:
            fileContent = f.read()
        lineArr = fileContent.split("\n")
        for i in range(len(lineArr)):
            lineArr[i] = lineArr[i].replace("\n", " ").replace("\\", f"{escapeTag}\\{escapeTag}\\").replace('"', f'{escapeTag}\\"').replace("'", f"{escapeTag}\\{escapeTag}'")
        result = f" {escapeTag}\\n ".join(lineArr)
        if minimize:
            result = " ".join(result.split())
        return code.format(result)

    def _COMMAND_LINE_REFPLACE(self, code, args):
        return args[0]

    def _parseLine(self, index, line):
        commentPosition = line.find(self._commentTag)
        if commentPosition == -1:
            if self._onLineCallback:
                return self._onLineCallback(line)
            return line
        macroStartIndex = line.find(self._macroStartTag)
        macroEndIndex = line.find(self._macroEndTag)
        if macroStartIndex == -1 or macroEndIndex == -1 or macroStartIndex >= macroEndIndex:
            if self._onLineCallback:
                return self._onLineCallback(line)
            return line
        code = line[:commentPosition]
        macro = line[macroStartIndex + len(self._macroEndTag):macroEndIndex]
        body = macro.split("|")
        self._assert(len(body) >= 1, f"invalid macro, line: {line}")
        command = body[0].strip()
        args = [arg.strip() for arg in body[1:]]
        if hasattr(self, f'_COMMAND_{command}'):
            return getattr(self, f'_COMMAND_{command}')(code, args)
        elif self._onMacroCallback:
            return self._onMacroCallback(code, command, *args)
        else:
            return line

    def _processBuild(self):
        self._assert(isinstance(self._inputFiles, list) and len(self._inputFiles) > 0, "input files are not defined")
        self._assert(isinstance(self._commentTag, str), "comment tag is not defined")

        self._print("reading files ...")
        for i, path in enumerate(self._inputFiles):
            self._assert(files.is_file(path), f"file not found: {path}")
            with open(path, 'r') as f:
                content = f.read()
            self._assert(len(content) > 0, "input files are empty")
            lineArr = content.split("\n")
            if py.is_text(self._headFormat):
                currName = self._inputNames[i]
                currDate = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                headInfo = self._headFormat.format(currName, currDate)
                self._lineArr.append("")
                self._lineArr.append(f"{self._commentTag} {headInfo}")
                self._lineArr.append("")
            for index, line in enumerate(lineArr):
                newLine = self._parseLine(index, line)
                if isinstance(newLine, list):
                    self._lineArr.extend(newLine)
                elif isinstance(newLine, str):
                    self._lineArr.append(newLine)

        self._print("creating target ...")
        html = "\n".join(self._lineArr)
        self._assert(self._outputFile is not None, "output path not found")
        with open(self._outputFile, 'w') as f:
            f.write(html)
        self._print("writing target succeeded!")
        return self



